packages/block/src/header.ts

```diff
@@ -117,6 +117,20 @@ export class BlockHeader {
         throw new Error('invalid header. baseFeePerGas should be provided')
       }
     }
+    if (opts.common && number !== undefined && timestamp !== undefined) {
+      if (opts.common.isActivatedEIP(4844) && excessDataGas === undefined) {
+        // There are situations where the common points to the wrong HF. Clone the common and
+        // set the HF by timestamp to ensure we are at the right fork.
+        // Note, for instance in blockchain/db/manager in `getHeader` we use the common of the manager
+        // This could thus point to the wrong HF (for instance, at the Shanghai -> Cancun fork block)
+        // The HF will be set to Cancun, but the parent block should point to Shanghai, but instead points to Cancun.
+        const clone = opts.common?.copy()
```

Comment:
I am not sure if I am doing this right here. The common could also be instantiated correctly in `blockchain/db/manager.ts`, but this adds boilerplate code..? Not sure how to proceed here?

Reply:
Ah, yes, this feels too hacky. We should very much try to avoid to add additional Common/HF hacks into the code.

Will give this some closer look.

Reply:
Update: yes, we should definitely look what is happening here and why the HF is not set correctly.

So the blockchain DB manager [getHeader()](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/blockchain/src/db/manager.ts#L153) method *does* various preparations to set the HF correctly and uses the `opts.hardforkByTTD` option and everything.

And in Block header constructor `timestamp` e.g. [is used](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/block/src/header.ts#L186) for HF setting.

Again: so something is off.

We should remove the code fix below and analyze this along the failing scenario which uncovered here, happy to help!

Reply:
yes from my understanding of the currrent code working, db/manager.ts sets the correct fork.
if opts have any hardforkby options set, the resulting header's common will have hf correctly set which can be used to validate against he value's expected length.

Reply:
`db/manager.ts` takes the common from `blockchain`, and the problem here is that if you put the Cancun fork block then the common is thus at Cancun, but if you retrieve the parent header, it should be Shanghai but it is Cancun (which then throws if you try to deserialize the RLP). 

Note that the Common / fork is not saved into the DB, only the RLP. (So upon typing this it would very make much sense to setup the right common in `db/manager.ts` and not here)